### PR TITLE
feat(application): implement newrelic_application resource

### PR DESCRIPTION
**Pull request by zlesnr**
resolves: #52

### CODE REVIEW

File: newrelic/resource_newrelic_application.go [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+	app, err := client.APM.GetApplication(userApp.ID)
+	if err != nil {
+		return err
+	}
+
+	log.Printf("[INFO] Read found New Relic application %+v\n\n\n", app)
+
+	flattenApplication(app, d)
+
+	return nil
```
Comment (sanderblue) [inline newrelic/resource_newrelic_application.go]: Should we return result of `flattenApplication(app, d)`, such as when an error occurs and is returned in that function?

File: newrelic/resource_newrelic_application_test.go [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckNewRelicApplicationDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccNewRelicApplicationConfig(),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckNewRelicApplicationExists(testAccExpectedApplicationName),
+					resource.TestCheckResourceAttr(
```
Comment (sanderblue) [inline newrelic/resource_newrelic_application_test.go]: Fairly certain the **import** test handles verifying these values so no need to include them.

cc: @ctrombley

Comment (sanderblue) [reply newrelic/resource_newrelic_application_test.go]: We can still keep `testAccCheckNewRelicApplicationExists` though. That's a more custom check.

File: newrelic/resource_newrelic_application_test.go [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+resource "newrelic_application" "app" {
+	name = "%s"
+	app_apdex_threshold = "0.9"
+	end_user_apdex_threshold = "0.8"
+	enable_real_user_monitoring = true
+}
+`, testAccExpectedApplicationName)
+}
+
+func testAccCheckNewRelicApplicationExists(n string) resource.TestCheckFunc {
```
Comment (sanderblue) [inline newrelic/resource_newrelic_application_test.go]: I think we want to make a call to the API via the client to ensure it exists. At least that's what we're doing in the other resources I believe. Should be relatively easy copy and paste and replace a few things.

File: newrelic/resource_newrelic_application.go (hunk 1 above)
Comment (zlesnr) [reply newrelic/resource_newrelic_application.go]: I've adjusted the `flattenApplication()` to return an error, and adjusted here to return.

File: newrelic/resource_newrelic_application_test.go (hunk 2 above)
Comment (zlesnr) [reply newrelic/resource_newrelic_application_test.go]: Have done.

File: newrelic/resource_newrelic_application_test.go (hunk 3 above)
Comment (zlesnr) [reply newrelic/resource_newrelic_application_test.go]: Have done.

File: newrelic/resource_newrelic_application.go [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+	_, err := client.APM.UpdateApplication(userApp.ID, updateParams)
+	if err != nil {
+		return err
+	}
+
+	return resourceNewRelicApplicationRead(d, meta)
+}
+
+func resourceNewRelicApplicationDelete(d *schema.ResourceData, meta interface{}) error {
```
Comment (ctrombley) [inline newrelic/resource_newrelic_application.go]: I'm not familiar with this pattern either, allowing delete to remove the resource from the state file but not actually deleting anything.  Is there a motivation behind this?

What is stopping us from actually calling the delete endpoint?

File: newrelic/resource_newrelic_application_test.go [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckNewRelicApplicationDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccNewRelicApplicationConfig(),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckNewRelicApplicationExists(testAccExpectedApplicationName),
+				),
+			},
```
Comment (ctrombley) [inline newrelic/resource_newrelic_application_test.go]: Can we add an update step here?

File: website/docs/d/application.html.markdown [hunk 6]
```diff
@@ -6,6 +6,44 @@ description: |-
   Looks up the information about an application in New Relic.
 ---
 
+# Resource: newrelic_application
```
[the remark this answers is not in the document]
Comment (zlesnr) [reply website/docs/d/application.html.markdown]: Ohh, I missed that bit in the organization.

File: newrelic/resource_newrelic_application.go [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+			},
+			"enable_real_user_monitoring": {
+				Type:     schema.TypeBool,
+				Required: true,
+			},
+		},
+	}
+}
+
+func resourceNewRelicApplicationCreate(d *schema.ResourceData, meta interface{}) error {
```
[the remark this answers is not in the document]
Comment (zlesnr) [reply newrelic/resource_newrelic_application.go]: There is an AWS resource that behaves similarly, where its created out of band, but then managed as part of the TF workflow.  I think it was the VPC gateway or maybe NAT gateway.

File: newrelic/resource_newrelic_application_test.go (hunk 5 above)
Comment (zlesnr) [reply newrelic/resource_newrelic_application_test.go]: Added.

File: website/docs/d/application.html.markdown (hunk 6 above)
[the remark this answers is not in the document]
Comment (zlesnr) [reply website/docs/d/application.html.markdown]: Moved.

File: newrelic/resource_newrelic_application.go (hunk 4 above)
Comment (zlesnr) [reply newrelic/resource_newrelic_application.go]: Handled out of band.

**Pull request merged**